openedx/core/djangoapps/course_roles/tests/test_api.py

```diff
@@ -0,0 +1,135 @@
+        )
+        self.course_1_key = CourseKey.from_string(str(self.course_1.id))
+        CourseOverview.load_from_module_store(self.course_1_key)
+        self.role_1 = Role.objects.create(name="test_role_1")
+        self.role_2 = Role.objects.create(name="test_role_2")
+        self.role_3 = Role.objects.create(name="test_role_3")
+        self.role_4 = Role.objects.create(name="test_role_4")
+        self.role_5 = Role.objects.create(name="test_role_5")
+        self.service = Service.objects.create(name="test_service")
+        self.role_1.services.add(self.service)
+        self.role_2.services.add(self.service)
+        self.role_3.services.add(self.service)
+        self.role_4.services.add(self.service)
+        self.role_5.services.add(self.service)
+        self.permission_1 = CourseRolesPermission.MANAGE_CONTENT
+        self.permission_2 = CourseRolesPermission.MANAGE_COURSE_SETTINGS
+        self.permission_3 = CourseRolesPermission.MANAGE_ADVANCED_SETTINGS
+        self.permission_4 = CourseRolesPermission.VIEW_ALL_CONTENT
+        self.permission_5 = CourseRolesPermission.VIEW_ONLY_LIVE_PUBLISHED_CONTENT
+        permission_1 = Permission.objects.create(name=self.permission_1.value.name)
+        permission_2 = Permission.objects.create(name=self.permission_2.value.name)
+        permission_3 = Permission.objects.create(name=self.permission_3.value.name)
+        permission_4 = Permission.objects.create(name=self.permission_4.value.name)
+        permission_5 = Permission.objects.create(name=self.permission_5.value.name)
+        self.role_1.permissions.add(permission_1)
+        self.role_2.permissions.add(permission_2)
+        self.role_3.permissions.add(permission_3)
+        self.role_4.permissions.add(permission_4)
+        self.role_5.permissions.add(permission_5)
+
+    def test_get_all_user_permissions_for_a_course_with_anonymus_user(self):
+        """
+        Test that get_all_user_permissions_for_a_course returns an empty set when the user is anonymous
+        """
+        assert not get_all_user_permissions_for_a_course(self.anonymous_user, self.course_1_key)
+
+    def test_get_all_user_permissions_for_a_course(self):
+        """
+        Test that get_all_user_permissions_for_a_course returns the correct permissions for the user and course
+        """
+        UserRole.objects.create(
+            user=self.user_1, role=self.role_1, course_id=self.course_1.id, org=self.organization_1
+        )
+        UserRole.objects.create(
+            user=self.user_2, role=self.role_4, course_id=self.course_1.id, org=self.organization_1
+        )
+        # Test that the correct permissions are returned for user_1
+        assert get_all_user_permissions_for_a_course(self.user_1, self.course_1_key) == {self.permission_1}
+        UserRole.objects.create(
+            user=self.user_1, role=self.role_2, course_id=self.course_1.id, org=self.organization_1
+        )
+        self.clear_caches()
```

Comment:
Is the cache clearing necessary here?

Reply:
If we're going to sometimes modify user roles in the same request where we're going to make queries on it, maybe we should also have an API call that will encapsulate `UserRole.objects.create()`, and let that new API call also remove the request cache entry for that user (i.e. do its own cache invalidation)?

Reply:
I agree, but we have not yet developed the role creation feature, we will cover it in a future PR and when we are going to do that we will have the clear cache planned